### PR TITLE
Surgery now goes slower before failing

### PR DESCRIPTION
**Pull request by PKPenguin321**
Based off of, and an alternative to, https://github.com/tgstation/tgstation/pull/46547
Credit to @jdawg1290 for a good bunch of the code.

## About The Pull Request

As in the original PR,
- Suboptimal tools and surfaces now delay surgery.
- Sterilizers now boost speed.

However, the slowness of a surgery step is now capped at 2x the original time the step would have taken. A step that takes 10 seconds would cap at 20 seconds, for example. Every second after that cap will instead be put into the odds of a failure.
So for example:
- Say a surgery takes 10 seconds by default, but you use a pen instead of a hemostat and do it on the floor instead of a surgery table.
- Now the surgery is slowed to 30 seconds, but actually it's capped at 20 seconds, so it takes 20 seconds.
- Those 10 seconds that were lost when we capped it to 20 seconds become your chance to fail, so now:
- The step takes 20 seconds and has a 10% chance to fail.

## Why It's Good For The Game

- As with the original PR, this will hopefully encourage the use of field surgery while limiting its effectiveness compared to a proper surgical suite
- As opposed to the original PR, this retains failed surgeries. Anything balanced around fails (lobotomy) is comparatively untouched, all the !!FUN!! of failing a surgery is untouched...
- Surgeries will never take "over 4 minutes" like in the original PR, which sounded awfully unfun to me
- ~~The hardest surgeries (like brainwashing) won't be doable in the absolute worst conditions (the fail prob with this formula will always be at or above 100% with that scale), which as many pointed out in the original PR was kind of dumb.~~ With permission from cobby, fail chance now caps at 99%, so there's always a 1% chance to pull off a surgery (good luck though).

## Changelog
:cl: PKPenguin321 and jdawg1290
balance: The chances of failing a surgery step have been changed. Now surgery will go slower the worse the surgery conditions are, capping at 2x the original surgery speed. Any time remaining after capping the surgery speed is converted to the odds of failing the surgery.
balance: Example: A surgery takes 10 seconds. In bad conditions, it now takes 30, but we cap it at 20. The 10 seconds lost to the cap become the chance to fail, so it also has a 10% chance of failing.
tweak: When you fail a surgery, the message you get will be a bit more informative as to your odds of succeeding if you try again.
/:cl:

Closes https://github.com/tgstation/tgstation/pull/46547

### CODE REVIEW

**Comment by PKPenguin321** [comment]
Here is what cobby said when I proposed this:
image

And some sample outputs:
image

**Comment by necromanceranne** [comment]
This is a good in between. Very cool.

**Comment by PKPenguin321** [comment]
Should I add a special message if the chance to fail is 100%, so people don't keep retrying surgeries in futile conditions? I feel like that would be a nice QoL thing.

**Comment by SkeletalElite** [comment]
> Should I add a special message if the chance to fail is 100%, so people don't keep retrying surgeries in futile conditions? I feel like that would be a nice QoL thing.

Maybe make the message differ depending on success chance so people can tell in game what kind of chances they have. 0%, <25%, <50%, <75% <100%, and 100% could all have different options.

**Comment by jdawg1290** [comment]
First, advanced surgeries require an operating computer which means you have a surgery table and these changes didn’t really affect you.
Second, the entire point of the change was to remove rng from surgery. Make it do 0-5 brute per step based on excess or something.

**Comment by PKPenguin321** [comment]
This does still remove RNG in a very significant number of circumstances. Both of the example images I gave in the case above are in the absolute worst scenarios (done on the floor with a pen and no sterilizers). A lot of the time you will only see the slowdown. You can also actively mitigate the RNG by using any sterilizer (such as alcohol), a better tool, or a bed/table. This allows RNG to exist without making it oppressive and un-mitigate-able.

**Comment by PKPenguin321** [comment]
>Maybe make the message differ depending on success chance so people can tell in game what kind of chances they have. 0%, <25%, <50%, <75%, <100%, and 100% could all have different text.

Latest commit does this

**Comment by Qustinnus** [comment]
Don't make it impossible, cap it at 1% so we can have that one guy that successfully does pen surgery

**Comment by PKPenguin321** [comment]
>Don't make it impossible, cap it at 1% so we can have that one guy that successfully does pen surgery

Should I? This is actually funny

**Comment by Qustinnus** [comment]
Hell yeah it won't have any real power game potential because it's 1%

**Comment by PKPenguin321** [comment]
It is done, cobby said it was fine

**Comment by PKPenguin321** [comment]
Any word on testmerges? I haven't really been following if it's been getting tested or not

**Comment by ExcessiveUseOfCobblestone** [comment]
I don't recall seeing slowdown on bagil, guess ill beg when i get home

**Comment by ExcessiveUseOfCobblestone** [comment]
I have only admin'd for the past week so thots mods?

**Comment by Donpedrito** [comment]
Mediborgs appear to be affected by the slowdown, despite normally having a 100% success chance on any non-dissection surgery in any conditions. Is this intended?

**Comment by PKPenguin321** [comment]
It would seem that borgs have an existing check to never fail surgeries, but aren't immune to surgery slowdown. This could be changed, but imo it seems fine. Use your roller bed instead of the floor if you don't want as much slowdown.

@ExcessiveUseOfCobblestone If you want it changed so borgs never have slowdown even in the worst conditions, let me know

**Comment by ExcessiveUseOfCobblestone** [comment]
that's fine.

I noticed stasis beds were making me fail a lot more than usual

**Comment by Mickyan** [comment]
I found a bunch of issues during testing, these may not all be caused by this PR but it's worth investigating

1. Stasis beds are definitely not giving the location modifier, using regular beds gives reliable success, stasis beds seem to be about the same as doing it on the ground
2. The message that's supposed to tell you what your failure chances are is not appearing
3. Failing the first step of tend wounds (scalpel) bugs the surgery, even if you later succeed you can only use the cautery to finish https://prnt.sc/pffd40.
In comparison, you should normally be seeing the two alternative steps like so https://prnt.sc/pff9d8

File: code/modules/surgery/helpers.dm [hunk 1]
```diff
@@ -114,7 +114,7 @@
 	else if(locate(/obj/structure/bed, T))
 		return 0.7
 	else
-		return 0.5
+		return 0.25
```
Comment (Mickyan) [inline code/modules/surgery/helpers.dm]: sure enough, missing a modifier for stasis beds here (should be 0.9)

**Comment by PKPenguin321** [comment]
>Stasis beds

Fixed in next commit

>The message that's supposed to tell you what your failure chances are is not appearing

25-49% fail chances have no extra message, and are just `You screw up!` as usual. If you can reproduce this while knowing precisely that your fail chance is out of that range, let me know. Also as far as I can tell there's no non-hacky way of applying the extra messages to non-standard steps, so if you were on a special step (any step that's not a regular incision/clamp/retract/etc, usually the last step of most surgeries), it not showing up is intended.

The third thing is weird. Not sure if I even understand what you're saying, but if I do, then I'm not sure that it's this PR that's causing the bug.

As for the floor surgery thing, try applying some alcohol or other sterilizers. The positive speed effect of those will help. If it's still too harsh, numbers can be tweaked later.

**Comment by PKPenguin321** [comment]
commit is not appearing in the PR for some reason wtf
let me try reopening

**Comment by Mickyan** [comment]
The only reason floor surgery is getting nerfed is the original author didn't even know it was already possible in the first place, it's not even in the changelog or no reason given for it, why keep it?

**Comment by PKPenguin321** [comment]
Floor surgery had no downsides before??

**Comment by Mickyan** [comment]
it had a failure modifier of 0.5, this PR changes it to 0.25 (see the lines I highlighted above)

**Comment by PKPenguin321** [comment]
Huh. Will look into code history in a bit when I'm at my computer. Thanks for pointing that out

**Comment by PKPenguin321** [comment]
Good catch. Don't know why that was changed, but it seems it is a holdover change from the original jdawg PR. I've undone it.

**Comment by ExcessiveUseOfCobblestone** [comment]
it takes it off automatically after every new commit (because that commit hasn't been reviewed)

File: code/modules/surgery/surgery_step.dm [hunk 2]
```diff
@@ -69,19 +70,28 @@
...
+		implement_speed_mod = implements[implement_type] / 100.0
 
-		if(implement_type)	//this means it isn't a require hand or any item step.
-			prob_chance = implements[implement_type]
-		prob_chance *= surgery.get_propability_multiplier()
+	speed_mod /= (get_location_modifier(target) * (1 + surgery.speed_modifier) * implement_speed_mod)
+	var/modded_time = time * speed_mod
+
+	fail_prob = min(max(0, modded_time - (time * 2)),99)//if modded_time > time * 2, then fail_prob = modded_time - time*2. starts at 0, caps at 99
+	modded_time = min(modded_time, time * 2)//also if that, then cap modded_time at time*2
```
Comment (PKPenguin321) [inline code/modules/surgery/surgery_step.dm]: To any future people who want to tweak these numbers, the * 2 on this line and the line above can be increased to make surgery go slower but fail less, and decreased to make surgery faster but fail more. I wouldn't recommend lowering it and anything above 3x seems excessive to me.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/surgery/surgery_step.dm]: should define 😏

Comment (PKPenguin321) [reply code/modules/surgery/surgery_step.dm]: >wasting a whole milisecond of compile time and a whole line of code to define a value that's used twice within a line from itself

😬

Comment (ExcessiveUseOfCobblestone) [reply code/modules/surgery/surgery_step.dm]: beauty comes at a price

**Pull request merged**